core/vm/contracts.go

```diff
@@ -630,27 +634,75 @@ func AddNewLock(blockContext BlockContext, statedb StateDB, toAddr common.Addres
+		statedb.SetState(internalContractAddress, toAddr.Hash(), common.BytesToHash([]byte{1}))
 		gasUsed += sstoreSetGas
 		stateGas += sstoreSetGas
-		currentLockHash = common.BytesToHash([]byte{1})
+		if gas.SubGas(sstoreSetGas) != nil {
+			return gasUsed, stateGas, ErrOutOfGas
+		}
+		// Set latest lock pointer also to 1
+		latestLockHash = common.BytesToHash([]byte{1})
 	}
-	currentLockNumber := new(big.Int).SetBytes(currentLockHash[:])
-	if !currentLockNumber.IsUint64() {
+	latestLockNumber := new(big.Int).SetBytes(latestLockHash[:])
+	if !latestLockNumber.IsUint64() {
 		return gasUsed, stateGas, errors.New("account has locked too many times, overflows uint64")
 	}
 
-	for i := int64(0); i < math.MaxInt64; i++ { // TODO: We should decide on a reasonable limit
-		// Ensure we have enough gas to complete this step entirely
-		requiredGas := coldSloadCost + sstoreSetGas + sstoreSetGas
+	key := toAddr.Bytes()
+	// Append latest lock to the key
+	key = binary.BigEndian.AppendUint64(key, latestLockNumber.Uint64())
+	key = append(key, byte(1)) // Set the 29th byte of the key to 1 for lockup
+	if len(key) > common.HashLength {
+		return gasUsed, stateGas, errors.New("lockup key is too long, math is broken")
+	}
+	lockHash := statedb.GetState(internalContractAddress, common.BytesToHash(key))
+	gasUsed += coldSloadCost
+	stateGas += coldSloadCost
+	if gas.SubGas(coldSloadCost) != nil {
+		// This contract does not revert. If the caller runs out of gas, we just stop
+		return gasUsed, stateGas, ErrOutOfGas
+	}
+	if (lockHash == common.Hash{}) {
+		// Lock doesn't exist, so add the new one here
+		statedb.SetState(internalContractAddress, common.BytesToHash(key), common.BytesToHash(lock.Bytes()))
+		gasUsed += sstoreSetGas
+		stateGas += sstoreSetGas
+		if gas.SubGas(sstoreSetGas) != nil {
+			return gasUsed, stateGas, ErrOutOfGas
+		}
+		key[28] = 0 // Set the 29th byte of the key to 0 for balance
+		statedb.SetState(internalContractAddress, common.BytesToHash(key), common.BytesToHash(balance.Bytes()))
+		gasUsed += sstoreSetGas
+		stateGas += sstoreSetGas
+		if gas.SubGas(sstoreSetGas) != nil {
+			return gasUsed, stateGas, ErrOutOfGas
+		}
+		// Increment the latest lock counter
+		latestLockNumber.Add(latestLockNumber, big1)
+		statedb.SetState(internalContractAddress, senderHash, common.BytesToHash(latestLockNumber.Bytes()))
+		gasUsed += sstoreSetGas
+		stateGas += sstoreSetGas
+		if gas.SubGas(sstoreSetGas) != nil {
+			return gasUsed, stateGas, ErrOutOfGas
+		}
+		// Addition of new lock successful
+		return gasUsed, stateGas, nil
+	} else { // This can only in the event of a bug (the latest lock number is wrong)
 		if gas.Gas() < requiredGas {
 			return gasUsed, stateGas, fmt.Errorf("insufficient gas to add new lock, required %d, got %d", requiredGas, gas.Gas())
 		}
+		log.Global.Infof("latest lock number is wrong for address %s: %s %s", toAddr.String(), senderHash, latestLockHash.String())
```

Comment:
IS this an info print or a warn or an error ?

Reply:
It's an error

Reply:
It's unlikely to ever hit and we use log.Global.Infof throughout the codebase